// gitlab-ci-variables-cli/bin/lib/gitlab-ci.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _bluebird = require('bluebird');

exports.default = gitlabCI;

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

var _urlParse = require('url-parse');

var _urlParse2 = _interopRequireDefault(_urlParse);

var _isPrimitive = require('is-primitive');

var _isPrimitive2 = _interopRequireDefault(_isPrimitive);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Provides utility functions to simplify interacting with a GitLab CI project through the API
 *
 * @param {string} url
 * @param {string} token
 *
 * @return {object} utility functions wrapped in an object
 */
function gitlabCI(url, token) {

  /**
   * Create project variable
   *
   * @param key
   * @param value
   *
   * @return {Promise<Object>} variable object
   */
  let createVariable = (() => {
    var _ref = (0, _bluebird.coroutine)(function* (key, value) {
      const response = yield (0, _axios2.default)({
        method: 'post',
        url: `${apiUrl}?${tokenQueryString}`,
        data: {
          key,
          value: serialiseValue(value)
        }
      });

      console.log(`Created new variable ${key} = ${JSON.stringify(value)}`);

      return response.data;
    });

    return function createVariable(_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();

  /**
   * Update project variable
   *
   * @param key
   * @param value
   *
   * @return {Promise<Object>} variable object
   */


  let updateVariable = (() => {
    var _ref2 = (0, _bluebird.coroutine)(function* (key, value) {
      const response = yield (0, _axios2.default)({
        method: 'put',
        url: `${apiUrl}/${key}?${tokenQueryString}`,
        data: {
          key,
          value: serialiseValue(value)
        }
      });

      console.log(`Updated variable ${key} = ${JSON.stringify(value)}`);

      return response.data;
    });

    return function updateVariable(_x3, _x4) {
      return _ref2.apply(this, arguments);
    };
  })();

  /**
   * Get all variables for project
   *
   * @return {Promise<Array>} array of variable objects
   */


  let listVariables = (() => {
    var _ref3 = (0, _bluebird.coroutine)(function* () {
      const response = yield _axios2.default.get(`${apiUrl}?${tokenQueryString}&${perPageQueryString}`);

      return response.data;
    });

    return function listVariables() {
      return _ref3.apply(this, arguments);
    };
  })();

  /**
   * Set project variables
   *
   * @param {Object} properties
   * @param forceUpdate if true, override existing values, otherwise ignore them
   *
   * @return {Promise<Array>} array of variable objects
   */


  let setVariables = (() => {
    var _ref4 = (0, _bluebird.coroutine)(function* (properties, forceUpdate) {
      if (!properties) {
        return null;
      }

      const existingKeys = (yield listVariables()).map(function (variable) {
        return variable.key;
      });
      const keysToSet = Object.keys(properties);

      const promises = keysToSet.map((() => {
        var _ref5 = (0, _bluebird.coroutine)(function* (key) {
          const value = properties[key];
          const keyExists = existingKeys.includes(key);

          if (keyExists && !forceUpdate) {
            console.log(`Skipped variable ${key}, already set for project.`);
            return undefined;
          }

          let variable;
          if (keyExists) {
            // Update variable
            variable = yield updateVariable(key, value);
          } else {
            // Create variable
            variable = yield createVariable(key, value);
          }

          return variable;
        });

        return function (_x7) {
          return _ref5.apply(this, arguments);
        };
      })());

      const variables = yield Promise.all(promises);

      return variables.filter(function (variable) {
        return variable;
      });
    });

    return function setVariables(_x5, _x6) {
      return _ref4.apply(this, arguments);
    };
  })();

  const parsedUrl = new _urlParse2.default(url);
  const perPageDefault = 100;

  // Construct project id by encoding namespace/projectName
  const projectId = parsedUrl.pathname.split('/').filter(x => x).join('%2F');

  const apiUrl = `${parsedUrl.origin}/api/v4/projects/${projectId}/variables`;
  const tokenQueryString = `private_token=${token}`;
  const perPageQueryString = `per_page=${perPageDefault}`;

  /**
   * Will serialise a value using `JSON.stringify` if it is not primative.
   *
   * @param {any} value
   * @return {string|number|boolean}
   */
  function serialiseValue(value) {
    if (!(0, _isPrimitive2.default)(value)) {
      return JSON.stringify(value);
    }

    return value;
  }

  return {
    createVariable,
    updateVariable,
    listVariables,
    setVariables
  };
}
module.exports = exports['default'];
